apps/prairielearn/src/lib/editors.js

```diff
@@ -56,425 +92,106 @@ class Editor {
-              job.fail(err);
-            } else {
-              job.succeed();
-            }
-          });
-        });
-      };
-
-      const _cleanupAfterCommit = (id) => {
-        debug(`Job id ${id} has failed (after git commit)`);
-        jobSequenceHasFailed = true;
-        _reset(_unlock, _unlock);
-      };
-
-      const _cleanupAfterWrite = (id) => {
-        debug(`Job id ${id} has failed (after write)`);
-        jobSequenceHasFailed = true;
-        _clean(_unlock, _unlock);
-      };
-
-      const _cleanup = (id) => {
-        debug(`Job id ${id} has failed`);
-        jobSequenceHasFailed = true;
-        _unlock();
-      };
-
-      const _unlock = () => {
-        debug(`${job_sequence_id}: _unlock`);
-        const jobOptions = {
-          course_id: options.course_id,
-          user_id: options.user_id,
-          authn_user_id: options.authn_user_id,
-          type: 'unlock',
-          description: 'Unlock',
-          job_sequence_id: job_sequence_id,
-          on_success: jobSequenceHasFailed ? _finishWithFailure : _finishWithSuccess,
-          on_error: _finishWithFailure,
-          no_job_sequence_update: true,
-        };
-        serverJobs.createJob(jobOptions, (err, job) => {
-          if (ERR(err, (e) => logger.error('Error in createJob()', e))) {
-            _finishWithFailure();
-            return;
-          }
 
-          namedLocks.releaseLock(courseLock, (err) => {
-            if (ERR(err, (e) => logger.error('Error in releaseLock()', e))) {
-              job.fail(err);
-            } else {
-              job.verbose(`Released lock`);
-              job.succeed();
-            }
-          });
-        });
-      };
-
-      const _finishWithSuccess = () => {
-        debug(`${job_sequence_id}: _finishWithSuccess`);
-        const jobOptions = {
-          course_id: options.course_id,
-          user_id: options.user_id,
-          authn_user_id: options.authn_user_id,
-          type: 'finish',
-          description: 'Finish job sequence',
-          job_sequence_id: job_sequence_id,
-          last_in_sequence: true,
-        };
-        serverJobs.createJob(jobOptions, (err, job) => {
-          if (ERR(err, (e) => logger.error('Error in createJob()', e))) {
-            _finishWithFailure();
-            return;
-          }
-
-          job.verbose('Finished with success');
-          job.succeed();
-          callback(null, job_sequence_id);
-        });
-      };
+              try {
+                await job.exec('git', ['push'], {
+                  cwd: this.course.path,
+                  env: gitEnv,
+                });
+              } catch (err) {
+                // In the original implementation, we would *only* reset the
+                // repo state. AFAICT, it should be safe to do both a clean
+                // and reset if the push fails?
+                // TODO: answer above question.
```

Comment:
👀

Reply:
It feels to me like we can always clean whenever we are resetting.

Reply:
Cool, I finalized that here: fbfcd93. I also used `finally` to reduce failure and success into a single case.